### PR TITLE
Indicium blog post

### PR DESCRIPTION
**Pull request by gsathya**

### CODE REVIEW

**Comment by mathiasbynens** [comment]
@RReverser @surma Are you on top of this?

**Comment by zeynepCankara** [comment]
@googlebot I consent.

**Review by camillobruni — APPROVED** [review]
LGTM

**Review by camillobruni — APPROVED** [review]

File: src/blog/system-analyzer.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+- an IC panel to get statistics about the IC events,
+- a Source panel to display Map/IC file positions on a script.
+
+![System Analyzer Overview](/_img/system-analyzer/system-analyzer-overview.png)
+
+![Group IC events by function name to get in depth information about the IC events associated with the `dotProduct`.](/_img/system-analyzer/case1_1.png)
+
+We are analyzing how the function `dotProduct` might be causing this performance difference. So we group IC events by functionName to get more in depth information about IC events associated with the `dotProduct` function.
+
+The first thing we notice is that we have two different IC state transitions recorded by the IC events in this function. One going from uninitialised to monomorphic and the other one going from monomorphic to polymorphic. As mentioned in the previously linked [blog post](https://mathiasbynens.be/notes/shapes-ics), polymorphic IC state indicates that now we are tracking more than one Map associated with `Point` objects.
```
Comment (rmcilroy) [inline src/blog/system-analyzer.md:116]: in the previously linked > in a previous

File: src/blog/system-analyzer.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+We want to know why we are creating multiple Map shapes for the same type of objects we toggle the info button about IC state to get more information about the Map addresses going from uninitialised to monomorphic.
```
Comment (rmcilroy) [inline src/blog/system-analyzer.md:118]: we toggle -> . To do so, we can

File: src/blog/system-analyzer.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+![The map transition tree associated with the monomorphic IC state.](/_img/system-analyzer/case1_2.png)
+
+![The map transition tree associated with the polymorphic IC state.](/_img/system-analyzer/case1_3.png)
+
+For the monomorphic IC state we can visualise the transition tree and see that we are only dynamically adding two properties `x`and `y` but when it comes to polymorphic IC state, we have a new Map containing three properties `isNegative`, `x` and `y`.
+
+![The Map panel communicates the file position information to highlight file positions on the Source panel.](/_img/system-analyzer/case1_4.png)
```
Comment (rmcilroy) [inline src/blog/system-analyzer.md:126]: The text in this image: "cause dynamically adding property" -> "causes the addition of a dynamically added property"

File: src/blog/system-analyzer.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+We click on the file position section of the Map panel to see where this `isNegative` property is added in the source code.
+
+So now the question being *how can we address the performance regression by using the insight we generated from the tool*?
```
Comment (rmcilroy) [inline src/blog/system-analyzer.md:130]: I'd drop this paragraph and just add " and can use this insight to address the performance regression" to the sentence above.

File: src/blog/system-analyzer.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+The minimal solution would be to always initialise the `isNegative` property. As a common advice we say that all instance properties should be initialised in the constructor.
```
Comment (rmcilroy) [inline src/blog/system-analyzer.md:132]: "As a common advice we say" -> "In general it is sound advice"

File: src/blog/system-analyzer.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+   this.y = y;
+ }
+
+ dotProduct(other) {
+   return this.x * other.x + this.y * other.y;
+ }
+}
+```
+
+If we execute the script again with the modified `Point` class, we see on the that execution of the two snippets we defined at the beginning of the case study perform very similarly.
```
Comment (rmcilroy) [inline src/blog/system-analyzer.md:150]: "we see on the that execution of the two snippets we defined at the beginning of the case study perform very similarly." -> "we see that the execution of the two snippets defined at the beginning of the case study perform very similarly."

File: src/blog/system-analyzer.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+![Performance analysis of snippets.](/_img/system-analyzer/final-program-performance.png)
```
Comment (rmcilroy) [inline src/blog/system-analyzer.md:152]: I'd drop this graph

File: src/blog/system-analyzer.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+![The map transition tree of the modified Point object.](/_img/system-analyzer/case2_1.png)
+
+Then we group IC events by `dotProduct` key in the IC panel. And now, we see that the polymorphic IC state is avoided as we are not creating multiple maps for the same type of objects.
```
Comment (rmcilroy) [inline src/blog/system-analyzer.md:156]: Move this above the image

Comment (rmcilroy) [reply src/blog/system-analyzer.md:156]: In fact, I'd just keep "In an updated trace, we see that the polymorphic IC state is avoided as we are not creating multiple maps for the same type of objects." at the end of the paragraph above.

File: src/blog/system-analyzer.md [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+### Map Panel
+
+The Map panel has two sub panels:
+
+1. Map details
+
+2. Map transitions
+
+The Map panel visualizes the transition trees of selected maps. The metadata of the selected map displayed through the map details sub-panel. A specific transition tree associated with a map address can be searched through using the provided interface. From the Stats sub-panel, which is above the Map transitions sub-panel, we can see the statistics about the properties causing map transitions and types of map events.
```
Comment (rmcilroy) [inline src/blog/system-analyzer.md:178]: searched through -> searched for

File: src/blog/system-analyzer.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+  - 'zeynep-cankara'
+date:
+tags:
+  - tools
+  - system-analyzer
+description: 'Indicium: V8 system analyzer tool to analyze Map/IC events.'
+---
+# Indicium: V8 system analyzer
+
+Past three months have been an awesome learning experience for me as I've joined the V8 team (Google London) as an intern, and have been  working on a new system analyzer tool called [*Indicium*](https://v8.dev/tools/system-analyzer).
```
Comment (rmcilroy) [inline src/blog/system-analyzer.md:14]: Past three -> The past three

File: src/blog/system-analyzer.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+This system analyzer is a unified web interface to trace, debug and analyse patterns of how Inline Caches (ICs) and Maps are created and modified in real-world applications.
+
+V8 already has a tracing infrastructure for [ICs](https://mathiasbynens.be/notes/shapes-ics) and [Maps](https://v8.dev/blog/fast-properties) which can process and analyse IC events using the [IC Explorer](https://v8.dev/tools//ic-explorer.html) and Map events using [Map Processor](https://v8.dev/tools/map-processor.html). Previous tools didn't allow us to analyze maps and ICs holistically and this is now possible with system analyzer .
```
Comment (rmcilroy) [inline src/blog/system-analyzer.md:18]: Previous -> However, previous

File: src/blog/system-analyzer.md (hunk 2 above)
Comment (gsathya) [reply src/blog/system-analyzer.md:118]: Done

File: src/blog/system-analyzer.md (hunk 4 above)
Comment (gsathya) [reply src/blog/system-analyzer.md:130]: Done

File: src/blog/system-analyzer.md (hunk 5 above)
Comment (gsathya) [reply src/blog/system-analyzer.md:132]: Done

File: src/blog/system-analyzer.md (hunk 6 above)
Comment (gsathya) [reply src/blog/system-analyzer.md:150]: Done

File: src/blog/system-analyzer.md (hunk 8 above)
Comment (gsathya) [reply src/blog/system-analyzer.md:156]: Done

File: src/blog/system-analyzer.md (hunk 9 above)
Comment (gsathya) [reply src/blog/system-analyzer.md:178]: Done

File: src/blog/system-analyzer.md (hunk 11 above)
Comment (gsathya) [reply src/blog/system-analyzer.md:18]: Done

File: src/blog/system-analyzer.md (hunk 7 above)
Comment (gsathya) [reply src/blog/system-analyzer.md:152]: Done

File: src/blog/system-analyzer.md (hunk 10 above)
Comment (gsathya) [reply src/blog/system-analyzer.md:14]: Done

File: src/blog/system-analyzer.md (hunk 3 above)
Comment (gsathya) [reply src/blog/system-analyzer.md:126]: Images can't be edited anymore :(

Gonna leave this is as such as I don't want to delete it

File: src/blog/system-analyzer.md (hunk 1 above)
Comment (gsathya) [reply src/blog/system-analyzer.md:116]: removed this line entirely

**Comment by RReverser** [comment]
Can you recompress images with https://squoosh.app/ or command-line OptiPNG / OxiPNG? E.g. trying case1_1.png got me from 194 KB to 167 KB, I imagine others will give similar savings.

**Comment by gsathya** [comment]
> Can you recompress images with https://squoosh.app/ or command-line OptiPNG / OxiPNG? E.g. trying case1_1.png got me from 194 KB to 167 KB, I imagine others will give similar savings.

```
** Processing: src/_img/system-analyzer/case1_1.png
954x706 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 167853 bytes
Input file size = 168372 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 167853

src/_img/system-analyzer/case1_1.png is already optimized.

** Processing: src/_img/system-analyzer/case1_2.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 154661 bytes
Input file size = 154766 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 154661

Output IDAT size = 154661 bytes (no change)
Output file size = 154718 bytes (48 bytes = 0.03% decrease)

** Processing: src/_img/system-analyzer/case1_3.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 166588 bytes
Input file size = 166705 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 166588

Output IDAT size = 166588 bytes (no change)
Output file size = 166645 bytes (60 bytes = 0.04% decrease)

** Processing: src/_img/system-analyzer/case1_4.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 167976 bytes
Input file size = 168093 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 167976

Output IDAT size = 167976 bytes (no change)
Output file size = 168033 bytes (60 bytes = 0.04% decrease)

** Processing: src/_img/system-analyzer/case2_1.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 163427 bytes
Input file size = 163532 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 163427

Output IDAT size = 163427 bytes (no change)
Output file size = 163484 bytes (48 bytes = 0.03% decrease)

** Processing: src/_img/system-analyzer/ic-panel2.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 138364 bytes
Input file size = 138469 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 138364

Output IDAT size = 138364 bytes (no change)
Output file size = 138421 bytes (48 bytes = 0.03% decrease)

** Processing: src/_img/system-analyzer/ic-panel3.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 98924 bytes
Input file size = 99017 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 98924

Output IDAT size = 98924 bytes (no change)
Output file size = 98981 bytes (36 bytes = 0.04% decrease)

** Processing: src/_img/system-analyzer/ic-panel4.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 101912 bytes
Input file size = 102005 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 101912

Output IDAT size = 101912 bytes (no change)
Output file size = 101969 bytes (36 bytes = 0.04% decrease)

** Processing: src/_img/system-analyzer/ic-panel.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 133895 bytes
Input file size = 134000 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 133895
  zc = 9  zm = 8  zs = 0  f = 4         IDAT size = 133812
  zc = 9  zm = 8  zs = 0  f = 5         IDAT size = 133493

Selecting parameters:
  zc = 9  zm = 8  zs = 0  f = 5         IDAT size = 133493

Output IDAT size = 133493 bytes (402 bytes decrease)
Output file size = 133550 bytes (450 bytes = 0.34% decrease)

** Processing: src/_img/system-analyzer/indicium-logo.png
122x145 pixels, 4x8 bits/pixel, RGB+alpha
Reducing image to 3x8 bits/pixel, RGB
Input IDAT size = 27109 bytes
Input file size = 27857 bytes

Trying:
  zc = 9  zm = 9  zs = 0  f = 0         IDAT size = 24144
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 24122
  zc = 9  zm = 9  zs = 0  f = 1         IDAT size = 22774
  zc = 9  zm = 9  zs = 1  f = 1         IDAT size = 21890
  zc = 9  zm = 9  zs = 1  f = 2         IDAT size = 21352
  zc = 9  zm = 9  zs = 1  f = 4         IDAT size = 21042
  zc = 9  zm = 9  zs = 1  f = 5         IDAT size = 20682

Selecting parameters:
  zc = 9  zm = 9  zs = 1  f = 5         IDAT size = 20682

Output IDAT size = 20682 bytes (6427 bytes decrease)
Output file size = 21418 bytes (6439 bytes = 23.11% decrease)

** Processing: src/_img/system-analyzer/initial-program-performance.png
600x371 pixels, 4x8 bits/pixel, RGB+alpha
Reducing image to 8 bits/pixel, 206 colors in palette
Input IDAT size = 8096 bytes
Input file size = 8153 bytes

Trying:
  zc = 9  zm = 9  zs = 0  f = 0         IDAT size = 3711
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 3711

Selecting parameters:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 3711

Output IDAT size = 3711 bytes (4385 bytes decrease)
Output file size = 4398 bytes (3755 bytes = 46.06% decrease)

** Processing: src/_img/system-analyzer/map-panel.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 157908 bytes
Input file size = 158013 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 157908

Output IDAT size = 157908 bytes (no change)
Output file size = 157965 bytes (48 bytes = 0.03% decrease)

** Processing: src/_img/system-analyzer/performance-graph.png
600x371 pixels, 4x8 bits/pixel, RGB+alpha
Reducing image to 3x8 bits/pixel, RGB
Input IDAT size = 10338 bytes
Input file size = 10395 bytes

Trying:
  zc = 9  zm = 9  zs = 0  f = 0         IDAT size = 9448
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 9448

Selecting parameters:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 9448

Output IDAT size = 9448 bytes (890 bytes decrease)
Output file size = 9505 bytes (890 bytes = 8.56% decrease)

** Processing: src/_img/system-analyzer/source-panel.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 92642 bytes
Input file size = 92723 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 92642

Output IDAT size = 92642 bytes (no change)
Output file size = 92699 bytes (24 bytes = 0.03% decrease)

** Processing: src/_img/system-analyzer/stats-panel.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 147396 bytes
Input file size = 147501 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 147396

Output IDAT size = 147396 bytes (no change)
Output file size = 147453 bytes (48 bytes = 0.03% decrease)

** Processing: src/_img/system-analyzer/system-analyzer-overview.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 177958 bytes
Input file size = 178075 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 177958

Output IDAT size = 177958 bytes (no change)
Output file size = 178015 bytes (60 bytes = 0.03% decrease)

** Processing: src/_img/system-analyzer/timeline-panel2.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 168643 bytes
Input file size = 168760 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 168643

Output IDAT size = 168643 bytes (no change)
Output file size = 168700 bytes (60 bytes = 0.04% decrease)

** Processing: src/_img/system-analyzer/timeline-panel.png
960x720 pixels, 4x8 bits/pixel, RGB+alpha
Input IDAT size = 169917 bytes
Input file size = 170034 bytes

Trying:
  zc = 9  zm = 8  zs = 0  f = 0         IDAT size = 169917

Output IDAT size = 169917 bytes (no change)
Output file size = 169974 bytes (60 bytes = 0.04% decrease)
```

**Review by RReverser — APPROVED** [review]
LGTM, thanks.

File: src/blog/system-analyzer.md [hunk 12]
```diff
@@ -0,0 +1,204 @@
+---
+title: 'Indicium: V8 runtime tracer tool'
+author: 'Zeynep Cankara ([@zcankara](https://twitter.com/zeynepcankara))'
+avatars:
+  - 'zeynep-cankara'
+date:
```
Comment (RReverser) [inline src/blog/system-analyzer.md:6]: @gsathya Please update the date to today.

**Comment by RReverser** [comment]
Wait, no, don't merge this yet! We have specific times when we publish :)

**Comment by gsathya** [comment]
> Wait, no, don't merge this yet! We have specific times when we publish :)

Oh no. Should I revert it?

**Comment by RReverser** [comment]
> Oh no. Should I revert it?

I'll do it.

**Comment by RReverser** [comment]
Done + restored the branch. Can you open a new PR? (I don't think it's possible to reopen)

**Comment by gsathya** [comment]
> Done + restored the branch. Can you open a new PR? (I don't think it's possible to reopen)

https://github.com/v8/v8.dev/pull/486

**Pull request merged**